File: benchmarks/drivers/python/bench_pipeline.py
"""Benchmark MongoCore pipeline batching at different batch sizes."""

import asyncio
import json
import os
import platform
import sys
import time
import statistics
from datetime import datetime, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "clients" / "python" / "src"))
from mongocore import MongoClient
from mongocore import ops

CONFIG = json.loads((Path(__file__).parent.parent / "common.json").read_text())
DATA_DIR = Path(__file__).parent.parent.parent / "data"
RESULTS_DIR = Path(__file__).parent.parent.parent / "results"
RESULTS_DIR.mkdir(exist_ok=True)

QUICK_MODE = "--quick" in sys.argv

WARMUP = 0 if QUICK_MODE else CONFIG["warmup_iterations"]["python"]
MIN_TIME = 0 if QUICK_MODE else CONFIG["min_time_secs"]
MAX_ITERS = 1 if QUICK_MODE else CONFIG["max_iterations"]
MAX_TIME = 5 if QUICK_MODE else CONFIG["max_time_secs"]
DB_NAME = CONFIG["database"]
ADDR = CONFIG["mongocore_address"]
SOCKET_PATH = CONFIG.get("mongocore_socket_path", "/tmp/mongocore.sock")

BATCH_SIZES = [100, 1000, 10000]
TOTAL_OPS = 10000

_ACTUAL_TRANSPORT = "tcp"


def get_system_info():
    return {
        "os": platform.system().lower(),
        "arch": platform.machine(),
        "cpus": os.cpu_count(),
        "mongocore_version": "0.6.0",
        "driver": "mongocore+python",
        "transport": _ACTUAL_TRANSPORT,
    }


async def run_benchmark(name, category, client, setup_fn, before_task_fn, task_fn, after_task_fn, teardown_fn, dataset_size_bytes, batch_size):
    """Run a benchmark following MongoDB spec methodology."""
    await setup_fn(client)

    for _ in range(WARMUP):
        await before_task_fn(client)
        await task_fn(client)
        await after_task_fn(client)

    times = []
    total_time = 0.0
    iteration = 0

    while total_time < MIN_TIME or iteration < 5:
        if iteration >= MAX_ITERS or total_time >= MAX_TIME:
            break

        await before_task_fn(client)
        start = time.perf_counter()
        await task_fn(client)
        elapsed = time.perf_counter() - start
        await after_task_fn(client)

        times.append(elapsed)
        total_time += elapsed
        iteration += 1

    await teardown_fn(client)

    times.sort()
    median = statistics.median(times)
    ops_per_sec = batch_size / median
    mb_per_sec = dataset_size_bytes / median / 1_000_000

    def percentile(data, pct):
        import math
        idx = max(0, math.ceil(len(data) * pct / 100) - 1)
        return data[min(idx, len(data) - 1)]

    result = {
        "benchmark": name,
        "category": category,
        "driver": "mongocore+python",
        "dataset_size_bytes": dataset_size_bytes,
        "batch_size": batch_size,
        "iterations": len(times),
        "total_time_secs": round(total_time, 3),
        "ops_per_sec": round(ops_per_sec, 1),
        "mb_per_sec": round(mb_per_sec, 3),
        "percentiles": {
            "p10": round(percentile(times, 10), 6),
            "p25": round(percentile(times, 25), 6),
            "p50": round(median, 6),
            "p75": round(percentile(times, 75), 6),
            "p90": round(percentile(times, 90), 6),
            "p95": round(percentile(times, 95), 6),
            "p99": round(percentile(times, 99), 6),
        },
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "system": get_system_info(),
    }
    print(f"  {name}: {ops_per_sec:.0f} ops/s, {mb_per_sec:.2f} MB/s ({len(times)} iters)")
    return result


async def main():
    print("=== MongoCore+Python Pipeline benchmarks ===")
    results = []

    small_doc = json.loads((DATA_DIR / "small_doc.json").read_text())
    tweet_doc = json.loads((DATA_DIR / "tweet.json").read_text())
    small_size = len(json.dumps(small_doc).encode())
    tweet_size = len(json.dumps(tweet_doc).encode())

    # Connect (try UDS first, fall back to TCP)
    client = None
    if os.path.exists(SOCKET_PATH):
        try:
            client = MongoClient(ADDR, socket_path=SOCKET_PATH, transport="grpc")
            await client.connect()
            await client.run_command("admin", {"hello": 1})
        except Exception:
            await client.close() if client else None
            client = None

    if client is None:
        client = MongoClient(ADDR, transport="grpc")
        await client.connect()
    else:
        global _ACTUAL_TRANSPORT
        _ACTUAL_TRANSPORT = "uds"

    for batch_size in BATCH_SIZES:
        calls_per_iter = TOTAL_OPS // batch_size

        # --- pipeline_run_command ---
        async def task_run_command(c, _bs=batch_size, _calls=calls_per_iter):
            for _ in range(_calls):
                operations = [ops.run_command(DB_NAME, {"hello": 1}) for _ in range(_bs)]
                await c.pipeline(*operations)

        results.append(await run_benchmark(
            f"pipeline_run_command_{batch_size}", "pipeline", client,
            setup_fn=lambda c: asyncio.sleep(0),
            before_task_fn=lambda c: asyncio.sleep(0),
            task_fn=task_run_command,
            after_task_fn=lambda c: asyncio.sleep(0),
            teardown_fn=lambda c: asyncio.sleep(0),
            dataset_size_bytes=TOTAL_OPS * 100,
            batch_size=TOTAL_OPS,
        ))

        # --- pipeline_insert_one_small ---
        async def before_insert(c):
            try:
                await c.run_command(DB_NAME, {"drop": "bench_pipeline_insert"})
            except Exception:
                pass

        async def task_insert_small(c, _bs=batch_size, _calls=calls_per_iter):
            from bson import ObjectId
            for _ in range(_calls):
                operations = [
                    ops.insert(DB_NAME, "bench_pipeline_insert", {**small_doc, "_id": str(ObjectId())})
                    for _ in range(_bs)
                ]
                await c.pipeline(*operations)

        results.append(await run_benchmark(
            f"pipeline_insert_one_small_{batch_size}", "pipeline", client,
            setup_fn=lambda c: asyncio.sleep(0),
            before_task_fn=before_insert,
            task_fn=task_insert_small,
            after_task_fn=lambda c: asyncio.sleep(0),
            teardown_fn=lambda c: asyncio.sleep(0),
            dataset_size_bytes=TOTAL_OPS * small_size,
            batch_size=TOTAL_OPS,
        ))

        # --- pipeline_find_one_by_id ---
        async def setup_find(c):
            try:
                await c.run_command(DB_NAME, {"drop": "bench_pipeline_find"})
            except Exception:
                pass
            coll = c[DB_NAME]["bench_pipeline_find"]
            await coll.insert_one({"_id": "bench_find_001", **tweet_doc})

        async def task_find_one(c, _bs=batch_size, _calls=calls_per_iter):
            for _ in range(_calls):
                operations = [
                    ops.find_one(DB_NAME, "bench_pipeline_find", {"_id": "bench_find_001"})
                    for _ in range(_bs)
                ]
                await c.pipeline(*operations)

        async def teardown_find(c):
            try:
                await c.run_command(DB_NAME, {"drop": "bench_pipeline_find"})
            except Exception:
                pass

        results.append(await run_benchmark(
            f"pipeline_find_one_by_id_{batch_size}", "pipeline", client,
            setup_fn=setup_find,
            before_task_fn=lambda c: asyncio.sleep(0),
            task_fn=task_find_one,
            after_task_fn=lambda c: asyncio.sleep(0),
            teardown_fn=teardown_find,
            dataset_size_bytes=TOTAL_OPS * tweet_size,
            batch_size=TOTAL_OPS,
        ))

    await client.close()

    output_path = RESULTS_DIR / "python_pipeline.json"
    with open(output_path, "w") as f:
        json.dump(results, f, indent=2)
    print(f"\nResults saved to {output_path}")


if __name__ == "__main__":
    asyncio.run(main())

File: benchmarks/drivers/typescript/bench_pipeline.ts
/**
 * Benchmark MongoCore pipeline batching at different batch sizes.
 * Uses the @mongocore/client library.
 */

import { readFileSync, mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';
import { performance } from 'perf_hooks';
import * as os from 'os';
import * as crypto from 'crypto';
import { MongoClient, ops } from '../../../clients/typescript/src';

const CONFIG = JSON.parse(readFileSync(join(__dirname, '..', 'common.json'), 'utf-8'));
const DATA_DIR = join(__dirname, '..', '..', 'data');
const RESULTS_DIR = join(__dirname, '..', '..', 'results');
mkdirSync(RESULTS_DIR, { recursive: true });

const QUICK_MODE = process.argv.includes('--quick');

const WARMUP = QUICK_MODE ? 0 : CONFIG.warmup_iterations.typescript;
const MIN_TIME = QUICK_MODE ? 0 : CONFIG.min_time_secs;
const MAX_ITERS = QUICK_MODE ? 1 : CONFIG.max_iterations;
const MAX_TIME = QUICK_MODE ? 5 : CONFIG.max_time_secs;
const DB_NAME = CONFIG.database;
const ADDR = CONFIG.mongocore_address;

const BATCH_SIZES = [100, 1000, 10000];
const TOTAL_OPS = 10000;

function newId(): string {
  return crypto.randomUUID().replace(/-/g, '').slice(0, 24);
}

interface BenchResult {
  benchmark: string;
  category: string;
  driver: string;
  dataset_size_bytes: number;
  batch_size: number;
  iterations: number;
  total_time_secs: number;
  ops_per_sec: number;
  mb_per_sec: number;
  percentiles: Record<string, number>;
  timestamp: string;
  system: Record<string, any>;
}

async function runBenchmark(
  name: string,
  category: string,
  setupFn: () => Promise<void>,
  beforeTaskFn: () => Promise<void>,
  taskFn: () => Promise<void>,
  afterTaskFn: () => Promise<void>,
  teardownFn: () => Promise<void>,
  datasetSizeBytes: number,
  batchSize: number,
): Promise<BenchResult> {
  await setupFn();

  for (let i = 0; i < WARMUP; i++) {
    await beforeTaskFn();
    await taskFn();
    await afterTaskFn();
  }

  const times: number[] = [];
  let totalTime = 0;
  let iteration = 0;

  while (totalTime < MIN_TIME || iteration < 5) {
    if (iteration >= MAX_ITERS || totalTime >= MAX_TIME) break;

    await beforeTaskFn();
    const start = performance.now();
    await taskFn();
    const elapsed = (performance.now() - start) / 1000;
    await afterTaskFn();

    times.push(elapsed);
    totalTime += elapsed;
    iteration++;
  }

  await teardownFn();

  times.sort((a, b) => a - b);
  const median = times[Math.floor(times.length / 2)];
  const opsPerSec = batchSize / median;
  const mbPerSec = datasetSizeBytes / median / 1_000_000;

  const pct = (p: number) => times[Math.min(Math.max(0, Math.ceil(times.length * p / 100) - 1), times.length - 1)];

  const result: BenchResult = {
    benchmark: name,
    category,
    driver: 'mongocore+typescript',
    dataset_size_bytes: datasetSizeBytes,
    batch_size: batchSize,
    iterations: times.length,
    total_time_secs: Math.round(totalTime * 1000) / 1000,
    ops_per_sec: Math.round(opsPerSec * 10) / 10,
    mb_per_sec: Math.round(mbPerSec * 1000) / 1000,
    percentiles: {
      p10: Math.round(pct(10) * 1000000) / 1000000,
      p25: Math.round(pct(25) * 1000000) / 1000000,
      p50: Math.round(median * 1000000) / 1000000,
      p75: Math.round(pct(75) * 1000000) / 1000000,
      p90: Math.round(pct(90) * 1000000) / 1000000,
      p95: Math.round(pct(95) * 1000000) / 1000000,
      p99: Math.round(pct(99) * 1000000) / 1000000,
    },
    timestamp: new Date().toISOString(),
    system: { os: os.platform(), arch: os.arch(), cpus: os.cpus().length, driver: 'mongocore+typescript', mongocore_version: '0.6.0', transport: 'tcp' },
  };

  console.log(`  ${name}: ${opsPerSec.toFixed(0)} ops/s, ${mbPerSec.toFixed(2)} MB/s (${times.length} iterations)`);
  return result;
}

async function main() {
  console.log('=== MongoCore+TypeScript Pipeline benchmarks ===');

  const client = new MongoClient(ADDR);
  await client.connect();
  const results: BenchResult[] = [];

  const smallDoc = JSON.parse(readFileSync(join(DATA_DIR, 'small_doc.json'), 'utf-8'));
  const tweetDoc = JSON.parse(readFileSync(join(DATA_DIR, 'tweet.json'), 'utf-8'));
  const smallSize = Buffer.byteLength(JSON.stringify(smallDoc));
  const tweetSize = Buffer.byteLength(JSON.stringify(tweetDoc));

  for (const batchSize of BATCH_SIZES) {
    const callsPerIter = TOTAL_OPS / batchSize;

    // --- pipeline_run_command ---
    results.push(await runBenchmark(
      `pipeline_run_command_${batchSize}`, 'pipeline',
      async () => {},
      async () => {},
      async () => {
        for (let c = 0; c < callsPerIter; c++) {
          const pipelineOps = Array.from({ length: batchSize }, () =>
            ops.runCommand(DB_NAME, { hello: 1 })
          );
          await client.pipeline(...pipelineOps);
        }
      },
      async () => {},
      async () => {},
      TOTAL_OPS * 100, TOTAL_OPS,
    ));

    // --- pipeline_insert_one_small ---
    results.push(await runBenchmark(
      `pipeline_insert_one_small_${batchSize}`, 'pipeline',
      async () => {},
      async () => {
        await client.runCommand(DB_NAME, { drop: 'bench_pipeline_insert_ts' }).catch(() => {});
      },
      async () => {
        for (let c = 0; c < callsPerIter; c++) {
          const pipelineOps = Array.from({ length: batchSize }, () =>
            ops.insert(DB_NAME, 'bench_pipeline_insert_ts', { ...smallDoc, _id: newId() })
          );
          await client.pipeline(...pipelineOps);
        }
      },
      async () => {},
      async () => {},
      TOTAL_OPS * smallSize, TOTAL_OPS,
    ));

    // --- pipeline_find_one_by_id ---
    const pipelineFindColl = client.db(DB_NAME).collection('bench_pipeline_find_ts');
    results.push(await runBenchmark(
      `pipeline_find_one_by_id_${batchSize}`, 'pipeline',
      async () => {
        await client.runCommand(DB_NAME, { drop: 'bench_pipeline_find_ts' }).catch(() => {});
        await pipelineFindColl.insertOne({ ...tweetDoc, _id: 'bench_find_001' });
      },
      async () => {},
      async () => {
        for (let c = 0; c < callsPerIter; c++) {
          const pipelineOps = Array.from({ length: batchSize }, () =>
            ops.findOne(DB_NAME, 'bench_pipeline_find_ts', { _id: 'bench_find_001' })
          );
          await client.pipeline(...pipelineOps);
        }
      },
      async () => {},
      async () => {
        await client.runCommand(DB_NAME, { drop: 'bench_pipeline_find_ts' }).catch(() => {});
      },
      TOTAL_OPS * tweetSize, TOTAL_OPS,
    ));
  }

  await client.close();

  const outputPath = join(RESULTS_DIR, 'typescript_pipeline.json');
  writeFileSync(outputPath, JSON.stringify(results, null, 2));
  console.log(`\nResults saved to ${outputPath}`);
}

main().catch((err) => { console.error(err); process.exit(1); });
